From b2d65942c297b879877c6ab0b72f718851ebcfd2 Mon Sep 17 00:00:00 2001
From: Dan Scally <djrscally@gmail.com>
Date: Wed, 9 Sep 2020 00:41:36 +0100
Subject: [PATCH 2/4] software node: fix refcount in graph_get_next_endpoint()

One of the problems we're having trying to build (using the changes you
attached here) a module to connect sensors to the cio2 infrastructure is
that we can't unload it cleanly. There seems to be a couple of reasons
for that; but one of them is that cio2_parse_firmware() in ipu3-cio2.c
ticks up the refcount for fwnode_handles of the ports for the CIO2 device
by calling software_node_graph_get_next_endpoint() once per _possible_
cio2 port; each time that happens it gets a reference to the port's
fwnode_handle but doesn't release it.

This isn't really a patch as such, since I don't think the changes you
attached are actually applied either upstream or in the media_tree git
(what are the plans in that regard, by the way? Will that patch be sent
upstream at some point?) so there's nowhere to apply it to, but I think
something like the below fixes it.

(patch from linux-media mailing list:
- https://lore.kernel.org/linux-media/2d4f1abb-c617-476a-1005-0ed91906a5f5@gmail.com/
  ("Re: [PATCH] media: ipu3: add a module to probe sensors via ACPI"))
[This patch is intended to be applied on top of the non-upstream swnode
patch ("software node: Add support for fwnode_graph* family of functions")
posted to the same thread of linux-media mailing list:
- https://lore.kernel.org/linux-media/20200526143110.GC3284396@kuha.fi.intel.com/]
[Converted spaces to tabs.]
Signed-off-by: Tsuchiya Yuto <kitakar@gmail.com>
---
 drivers/base/swnode.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/base/swnode.c b/drivers/base/swnode.c
index 9955a80e442c..da353a1c1e3e 100644
--- a/drivers/base/swnode.c
+++ b/drivers/base/swnode.c
@@ -580,6 +580,8 @@ software_node_graph_get_next_endpoint(const struct fwnode_handle *fwnode,
 		fwnode_handle_put(old);
 		if (endpoint)
 			break;
+		else
+			fwnode_handle_put(port);
 	}
 
 	fwnode_handle_put(port);
-- 
2.28.0

